File: 3/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
	const number int = 600851475143

	largest := 0

	for i := 3; i <= sqrt(number); i += 2 {
		if number%i == 0 && isPrime(i) {
			largest = i
		}
	}

	fmt.Println(largest)
}

func isPrime(num int) bool {
	for i := 3; i <= sqrt(num); i += 2 {
		if num%i == 0 {
			return false
		}
	}

	return true
}

func sqrt(i int) int {
	return int(math.Sqrt(float64(i)))
}

File: 3/main.hs
import Data.List

num = 600851475143

main = print $ find primeFactor [start, start - 1 .. 0]
	where
		start = sqrt' num
		primeFactor x = num `mod` x == 0 && isPrime x

isPrime :: Integer -> Bool
isPrime x = not $ any ((== 0) . (mod x)) $ [2..end]
	where end = sqrt' x

sqrt' :: Integer -> Integer
sqrt' = ceiling . sqrt . fromIntegral
